### PR TITLE
KEYCLOAK-18146 Search for clients by client attribute when doing saml…

### PR DESCRIPTION
**Pull request by mhajas**
… artifact resolution

### CODE REVIEW

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            }
+
+        } catch (Exception e) {
+            throw new CustomChangeException(getTaskId() + ": Exception when extracting data from previous version", e);
+        }
+    }
+
+    private boolean hasSamlAttributeIdentifier(String idOfClient) throws CustomChangeException {
+        try (PreparedStatement statement = jdbcConnection.prepareStatement("SELECT NAME FROM " + getTableName("CLIENT_ATTRIBUTES") +
+                " WHERE CLIENT_ID=? AND NAME=?")) {
```
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java:81]:
```suggestion
                " WHERE CLIENT_ID=? AND CLIENT_ID=?")) {
```

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+public class JpaUpdate14_0_0_MigrateSamlArtifactAttribute extends CustomKeycloakTask {
+
+    private static final String SAML_ARTIFACT_BINDING_IDENTIFIER = "saml.artifact.binding.identifier";
+
+    private final Map<String, String> clientIds = new HashMap<>();
+
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        extractClientsData("SELECT ID, CLIENT_ID FROM " + getTableName("CLIENT") + " " +
+                "WHERE PROTOCOL='saml'");
```
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java:39]:
```suggestion
        extractClientsData("SELECT c.ID, c.CLIENT_ID FROM " + getTableName("CLIENT") + " c " +
                " LEFT JOIN " + getTableName("CLIENT_ATTRIBUTES") + " ca ON c.CLIENT_ID = ca.CLIENT_ID AND ca.name = '" + SAML_ARTIFACT_BINDING_IDENTIFIER + "' " +
                "WHERE c.PROTOCOL='saml' AND ca.name IS NULL");
```

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            statement.setString(1, idOfClient);
+            statement.setString(2, SAML_ARTIFACT_BINDING_IDENTIFIER);
+            try (ResultSet rs = statement.executeQuery()) {
+                return rs.next(); // The client has saml attribute
+            }
+        } catch (Exception e) {
+            throw new CustomChangeException(getTaskId() + ": Exception when extracting data from previous version", e);
+        }
+    }
+
```
Comment (hmlnarik) [inline model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java:91]: This method can be removed if the query above is used.

File: server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java [hunk 4]
```diff
@@ -1406,6 +1408,11 @@ private static ClientModel createClient(KeycloakSession session, RealmModel real
             }
         }
 
+        if ("saml".equals(resourceRep.getProtocol())
```
Comment (hmlnarik) [inline server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java:1411]: Prefer `SamlProtocol.LOGIN_PROTOCOL` (other places as well)

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java [hunk 5]
```diff
@@ -87,6 +89,7 @@ public static ClientRepresentation createClientRep(String clientId, String proto
         attributes.put(SAML_SIGNATURE_ALGORITHM, "RSA_SHA256");
         attributes.put(SAML_FORCE_NAME_ID_FORMAT, "false");
         attributes.put(SAML_NAME_ID_FORMAT, "username");
+        attributes.put(SamlConfigAttributes.SAML_ARTIFACT_BINDING_IDENTIFIER, ArtifactBindingUtils.computeArtifactBindingIdentifierString("saml"));
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java:92]: Is this line really necessary? `SAML_ARTIFACT_BINDING_IDENTIFIER` should be automatically set on server when not present

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Overall looks good to me @mhajas! I believe the only thing that needs fixing is the migration.

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java (hunk 5 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java:92]: If I understand the test correctly, these attributes are just checked after the creation, they are not set when creating the client.

File: server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java (hunk 4 above)
Comment (mhajas) [reply server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java:1411]: `SamlProtocol` is part of services and `RepresentationToModel` is in `server-spi-private` so it is not available there. I can create some constants class in `server-spi-private` and then reuse it everywhere, but I am not sure this is what we want.

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java (hunk 5 above)
Comment (hmlnarik) [reply testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java:92]: It seems they are used also for creation, see e.g. `org.keycloak.testsuite.console.clients.ClientSettingsTest.createSAML()`
This line can stay here if there is another test (e.g. in base testsuite) that tests the creation of SAML client without this attribute and then checks that the attribute has been added by keycloak server when reading the client back.

File: server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java (hunk 4 above)
Comment (hmlnarik) [reply server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java:1411]: Ah, that makes perfect sense :)

File: testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java (hunk 5 above)
Comment (mhajas) [reply testsuite/integration-arquillian/tests/other/console/src/test/java/org/keycloak/testsuite/console/clients/AbstractClientTest.java:92]: From `newClient` `ClientRepresentation` only `clientId` and `protocol` is used for creation, see [this](https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/other/console/src/main/java/org/keycloak/testsuite/console/page/clients/CreateClientForm.java#L27-L31).

File: model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java (hunk 2 above)
Comment (mhajas) [reply model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate14_0_0_MigrateSamlArtifactAttribute.java:39]: Thanks, that is nice improvement. I will add it.

**Comment by mhajas** [comment]
@hmlnarik @vramik I updated the PR based on your feedback. I have also run Jenkins again, I will post the results once it is finished. 

Here are previous results:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/449/
https://keycloak-jenkins.com/job/universal-test-pipeline-adapters/396/

I fixed the failures of `JsonFileImport***MigrationTest` that was caused by my changes. I cannot identify anything else, but if you see anything else just let me know.

**Comment by mhajas** [comment]
Last Jenkins runs:
https://master-jenkins.com/job/universal-test-pipeline-server/454
https://master-jenkins.com/job/universal-test-pipeline-adapters/403/

**Review by hmlnarik — APPROVED** [review]
Thank you for the PR!

**Pull request merged**